metrics/scraper.go

```diff
@@ -13,72 +13,186 @@ import (
 // Scraper is an interface for scraping metrics data.
 type Scraper interface {
 	GetAverageCPUUtilizationByWorkload(namespace,
+		workload string,
+		start time.Time,
+		end time.Time,
+		step time.Duration) ([]float64, error)
+
+	GetCPUUtilizationBreachDataPoints(namespace,
 		workloadType,
-		workload string, start time.Time) ([]float64, error)
+		workload string,
+		redLineUtilization float64,
+		start time.Time,
+		end time.Time,
+		step time.Duration) ([]float64, error)
 }
 
 // PrometheusScraper is a Scraper implementation that scrapes metrics data from Prometheus.
 type PrometheusScraper struct {
-	api v1.API
+	api            v1.API
+	metricRegistry *MetricRegistry
+}
+
+type MetricRegistry struct {
+	utilizationMetric     string
+	podOwnerMetric        string
+	resourceLimitMetric   string
+	readyReplicasMetric   string
+	replicaSetOwnerMetric string
+	hpaMaxReplicasMetric  string
+	hpaOwnerInfoMetric    string
 }
 
 // NewPrometheusScraper returns a new PrometheusScraper instance.
 
 func NewPrometheusScraper(apiURL string) (*PrometheusScraper, error) {
+
 	client, err := api.NewClient(api.Config{
 		Address: apiURL,
 	})
+
 	if err != nil {
 		return nil, fmt.Errorf("error creating Prometheus client: %v", err)
 	}
+	cpuUtilizationMetric := "node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate"
+	podOwnerMetric := "namespace_workload_pod:kube_pod_owner:relabel"
+	resourceLimitMetric := "cluster:namespace:pod_cpu:active:kube_pod_container_resource_limits"
+	readyReplicasMetric := "kube_replicaset_status_ready_replicas"
+	replicaSetOwnerMetric := "kube_replicaset_owner"
+	hpaMaxReplicasMetric := "kube_horizontalpodautoscaler_spec_max_replicas"
+	hpaOwnerInfoMetric := "kube_horizontalpodautoscaler_info"
+
+	return &PrometheusScraper{api: v1.NewAPI(client),
+			metricRegistry: &MetricRegistry{utilizationMetric: cpuUtilizationMetric,
+				podOwnerMetric:        podOwnerMetric,
+				resourceLimitMetric:   resourceLimitMetric,
+				readyReplicasMetric:   readyReplicasMetric,
+				replicaSetOwnerMetric: replicaSetOwnerMetric,
+				hpaMaxReplicasMetric:  hpaMaxReplicasMetric,
+				hpaOwnerInfoMetric:    hpaOwnerInfoMetric,
+			}},
+		nil
+}
 
-	promAPI := v1.NewAPI(client)
+// GetAverageCPUUtilizationByWorkload returns the average CPU utilization for the given workload type and name in the
+// specified namespace, in the given time range.
+func (ps *PrometheusScraper) GetAverageCPUUtilizationByWorkload(namespace string,
+	workload string,
```

Comment:
This is currently implemented for only workload type 'deployment'?

Reply:
namespace_workload_pod:kube_pod_owner:relabel relabels the workload type to deployment for both rollouts and deployments. This fn will work for both rollout and deployment